### PR TITLE
[SPARK-42559][CONNECT] Implement DataFrameNaFunctions

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to implement DataFrameNaFunctions.

### Why are the changes needed?
API coverage.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Add new UT.

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+ */
+final class DataFrameNaFunctions private[sql](df: DataFrame) {
+
+  /**
+   * Returns a new `DataFrame` that drops rows containing any null or NaN values.
+   *
+   * @since 3.4.0
+   */
+  def drop(): DataFrame = {
+    df.sparkSession.newDataset { builder =>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:42]: I think you need to update your branch. `newDataset` takes an encoder now. You should use `newDataFrame`.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+import org.apache.spark.connect.proto.Expression.{Literal => GLiteral}
+import org.apache.spark.connect.proto.NAReplace
+import org.apache.spark.connect.proto.NAReplace.Replacement
+
+/**
+ * Functionality for working with missing data in `DataFrame`s.
+ *
+ * @since 3.4.0
+ */
+final class DataFrameNaFunctions private[sql](df: DataFrame) {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:34]: You don't need to pass the DataFrame here. Just pass the root relation.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala (hunk 1 above)
Comment (panbingkun) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:42]: OK

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala (hunk 2 above)
Comment (panbingkun) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:34]: ok

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+   *
+   * If `how` is "any", then drop rows containing any null or NaN values.
+   * If `how` is "all", then drop rows only if every column is null or NaN for that row.
+   *
+   * @since 3.4.0
+   */
+  def drop(how: String): DataFrame = {
+    sparkSession.newDataFrame { builder =>
+      val dropNaBuilder = builder.getDropNaBuilder.setInput(root)
+      how.toLowerCase(Locale.ROOT) match {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:58]: Can you put this in a separate method to reduce duplication?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+ */
+final class DataFrameNaFunctions private[sql](sparkSession: SparkSession, root: Relation) {
+
+  /**
+   * Returns a new `DataFrame` that drops rows containing any null or NaN values.
+   *
+   * @since 3.4.0
+   */
+  def drop(): DataFrame = {
+    sparkSession.newDataFrame { builder =>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:42]: Is there a way we can reduce duplication. For example:
```scala
private def buildDropDataFrame(cols: Seq[String], minNonNulls: Option[Int]): DataFrame = {
  sparkSession.newDataFrame { builder =>
    val dropNaBuilder = builder.getDropNaBuilder.setInput(root).addAllCols(cols.asJava)
    minNonNulls.foreach(builder.setMinNonNulls)
}
```

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:42]: The same applies for fill.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala (hunk 3 above)
Comment (panbingkun) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala]: Done

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala (hunk 4 above)
Comment (panbingkun) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala]: Done

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+  /**
+   * Returns a new `DataFrame` that drops rows containing null or NaN values.
+   *
+   * If `how` is "any", then drop rows containing any null or NaN values. If `how` is "all", then
+   * drop rows only if every column is null or NaN for that row.
+   *
+   * @since 3.4.0
+   */
+  def drop(how: String): DataFrame = {
+    val minNonNulls = how.toLowerCase(Locale.ROOT) match {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:53]: please put this in a helper function and reuse it here and in `def drop(how: String, cols: Seq[String]): DataFrame`.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+  def fill(value: String, cols: Array[String]): DataFrame = fill(value, cols.toSeq)
+
+  /**
+   * (Scala-specific) Returns a new `DataFrame` that replaces null values in specified string
+   * columns. If a specified column is not a string column, it is ignored.
+   *
+   * @since 3.4.0
+   */
+  def fill(value: String, cols: Seq[String]): DataFrame = {
+    buildFillDataFrame(Some(cols), convertToLiteral(value, StringType))
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:222]: AFAICT null is not allowed as a fill value, a null will fall through to the default in this code: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala#L517-L523

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+          Replacement
+            .newBuilder()
+            .setOldValue(convertToLiteral(oldValue, StringType))
+            .setNewValue(convertToLiteral(newValue, StringType))
+            .build()
+      }
+    }
+  }
+
+  private def convertToLiteral(v: Any, dataType: DataType): GLiteral = dataType match {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:473]: You don't need this. Just use `functions.lit(v).expr`. I looked at the SparkConnectPlanner and nulls handled properly in there.

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala:473]: This should also significantly simplify the code in `replacementMap` and `fillMap`.

Comment (panbingkun) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala]: OK, I have deleted convertToLiteral method

**Comment by hvanhovell** [comment]
@panbingkun can you update your PR.

**Comment by hvanhovell** [comment]
@panbingkun can you update the CompatibilitySuite?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionSuite.scala [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.sql.connect.client.util.QueryTest
+
+class DataFrameNaFunctionSuite extends QueryTest {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionSuite.scala:24]: Is this a line for line copy of the original test?

**Comment by panbingkun** [comment]
> @panbingkun can you update the CompatibilitySuite?

Done

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/QueryTest.scala [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+        b.keys.find(bKey => compare(aKey, bKey)).exists(bKey => compare(a(aKey), b(bKey)))
+      }
+    case (a: Iterable[_], b: Iterable[_]) =>
+      a.size == b.size && a.zip(b).forall { case (l, r) => compare(l, r) }
+    case (a: Product, b: Product) =>
+      compare(a.productIterator.toSeq, b.productIterator.toSeq)
+    case (a: Row, b: Row) =>
+      compare(a.toSeq, b.toSeq)
+    // 0.0 == -0.0, turn float/double to bits before comparison, to distinguish 0.0 and -0.0.
+    case (a: Double, b: Double) =>
```
Comment (panbingkun) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/QueryTest.scala:193]: We need it to compare the data of row.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 10]
```diff
@@ -134,6 +135,9 @@ object CheckConnectJvmClientCompatibility {
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.DataFrameReader.jdbc"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.DataFrameWriter.jdbc"),
 
+      // DataFrameNaFunctions
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.DataFrameNaFunctions.this"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:139]: Nice

**Comment by hvanhovell** [comment]
LGTM

**Comment by hvanhovell** [comment]
Merging

**Review by hvanhovell — APPROVED** [review]
LGTM

**Pull request closed without merging**